### PR TITLE
Create preview refactoring

### PR DESCRIPTION
**Pull request by baldummy**
We need to have a way to create preview environments based on git hashes instead of pull requests. Because of this the existing commands create-preview and delete-preview were refactored. They are now independent of pull requests. The new commands to use for pull request preview environments are now create-pr-preview and delete-pr-preview.

### CODE REVIEW

File: mkdocs.yml [hunk 1]
```diff
@@ -15,6 +15,7 @@ nav:
   - CLI Commands:
     - add-pr-comment: commands/add-pr-comment.md
     - create-preview: commands/create-preview.md
+    - create-pr-preview: commands/create-pr-preview.md
     - delete-preview: commands/delete-preview.md
```
Comment (christiansiegel) [inline mkdocs.yml:19]: `delete-pr-preview` missing

File: tests/test_cliparser.py [hunk 2]
```diff
@@ -8,23 +8,23 @@
 from gitopscli.cliparser import create_cli
 
 EXPECTED_GITOPSCLI_HELP = """\
-usage: gitopscli [-h]
-                 {deploy,sync-apps,add-pr-comment,create-preview,delete-preview,version}
-                 ...
+usage: gitopscli [-h] {deploy,sync-apps,add-pr-comment,create-preview-common,create-pr-preview,delete-preview,delete-pr-preview,version} ...
```
Comment (christiansiegel) [inline tests/test_cliparser.py:11]: `create-preview-common`? but I guess this test still fails..

File: gitopscli/commands/create_pr_preview.py [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+
+
+def __merge_pullrequest(branch, pull_request, root_git):
+    root_git.merge_pull_request(pull_request)
+    logging.info("Pull request merged")
+    root_git.delete_branch(branch)
+    logging.info("Branch '%s' deleted", branch)
+
+
+def __create_tmp_dir():
```
Comment (christiansiegel) [inline gitopscli/commands/create_pr_preview.py:120]: This is now available as a common function. You should merge the master branch :wink: 
https://github.com/baloise/gitopscli/commit/9b4dff2e29f7cc5690f2e496a2036550aca2678a

File: gitopscli/commands/create_preview.py [hunk 4]
```diff
@@ -71,24 +65,20 @@ def create_preview_command(
...
-
+        config_branch = f"master"
         preview_template_folder_name = ".preview-templates/" + gitops_config.application_name
         if os.path.isdir(root_git.get_full_file_path(preview_template_folder_name)):
             logging.info("Using the preview template folder: %s", preview_template_folder_name)
         else:
             raise GitOpsException(f"The preview template folder does not exist: {preview_template_folder_name}")
 
-        new_preview_folder_name = gitops_config.application_name + "-" + shortened_branch_hash + "-preview"
+        new_preview_folder_name = gitops_config.application_name + "-" + preview_id + "-preview"
```
Comment (christiansiegel) [inline gitopscli/commands/create_preview.py:75]: The preview id should be validated. the preview folder must conform to the k8s namespace name restrictions (I think it was max 63 chars, [a-z0-9] and dashes)

File: gitopscli/commands/create_preview.py [hunk 5]
```diff
@@ -13,21 +12,20 @@
...
     git_email,
-    create_pr,
-    auto_merge,
     organisation,
     repository_name,
     git_provider,
     git_provider_url,
+    git_hash,
+    preview_id,
+    deployment_replaced=None,
```
Comment (christiansiegel) [inline gitopscli/commands/create_preview.py:25]: I would rename those to `deployment_replaced_callback` etc. for clarity

File: gitopscli/commands/create_pr_preview.py [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+            deployment_exist(parent_id, pr_id, pr_branch),
+            deployment_new(parent_id, pr_id, pr_branch)
+        )
+    finally:
+        shutil.rmtree(apps_tmp_dir, ignore_errors=True)
+        shutil.rmtree(root_tmp_dir, ignore_errors=True)
+
+
+def deployment_replaced(parent_id, pr_id):
+    def inner_func(apps_git, new_image_tag):
```
Comment (christiansiegel) [inline gitopscli/commands/create_pr_preview.py:71]: I would rename the outer function to `create_deployment_replaced_callback`and the inner to `deployment_replaced_callback`/`callback` for clarity.

File: gitopscli/commands/create_preview.py [hunk 7]
```diff
@@ -107,33 +98,27 @@ def create_preview_command(
                 replacement,
```
Comment (christiansiegel) [inline gitopscli/commands/create_preview.py:98]: (GitHub doesn't allow to comment unchanged lines)
the `new_image_tag = apps_git.get_last_commit_hash()` will not work anymore. You should use the `git_hash` passed to the command.

File: gitopscli/commands/create_preview.py [hunk 8]
```diff
@@ -107,33 +98,27 @@ def create_preview_command(
...
                 root_git,
                 route_host,
-                shortened_branch_hash,
+                preview_id,
                 value_replaced,
             )
         if not value_replaced:
-            __no_deployment_needed(apps_git, new_image_tag, parent_id, pr_id)
+            if deployment_replaced_callback:
+                deployment_replaced_callback(apps_git, new_image_tag)
```
Comment (christiansiegel) [inline gitopscli/commands/create_preview.py:106]: IMO the naming makes no sense. Shouldn't it be "deployment_already_up_to_date" or at least "deployment_NOT_replaced"?

File: gitopscli/commands/delete_pr_preview.py [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+            git_email,
+            organisation,
+            repository_name,
+            git_provider,
+            git_provider_url,
+            apps_tmp_dir,
+        )
+
+        apps_git.checkout(branch)
+        preview_id = hashlib.sha256(branch.encode("utf-8")).hexdigest()[:8]
```
Comment (christiansiegel) [inline gitopscli/commands/delete_pr_preview.py:44]: again: branch eventually hashed twice

File: gitopscli/commands/delete_preview.py [hunk 10]
```diff
@@ -63,49 +60,29 @@ def delete_preview_command(
...
-    logging.info("Pull request created: %s", {root_git.get_pull_request_url(pull_request)})
-    return pull_request
 
 
-def __merge_pullrequest(branch, pull_request, root_git):
-    root_git.merge_pull_request(pull_request)
-    logging.info("Pull request merged")
-    root_git.delete_branch(branch)
-    logging.info("Branch '%s' deleted", branch)
+def __create_tmp_dir():
```
Comment (christiansiegel) [inline gitopscli/commands/delete_preview.py:84]: unused

File: gitopscli/commands/create_preview.py (hunk 8 above)
Comment (m4rc0z) [reply gitopscli/commands/create_preview.py:106]: you are right

File: gitopscli/commands/create_preview.py [hunk 11]
```diff
@@ -72,68 +66,58 @@ def create_preview_command(
...
             route_host, value_replaced = __replace_value(
                 gitops_config,
-                new_image_tag,
+                git_hash,
                 new_preview_folder_name,
                 replacement,
                 root_git,
                 route_host,
-                shortened_branch_hash,
+                preview_id,
```
Comment (christiansiegel) [inline gitopscli/commands/create_preview.py:100]:
```suggestion
                hashed_preview_id,
```

File: gitopscli/commands/create_pr_preview.py [hunk 12]
```diff
@@ -0,0 +1,103 @@
...
+            git_provider,
+            git_provider_url,
+            apps_tmp_dir,
+        )
+
+        pr_branch = apps_git.get_pull_request_branch(pr_id)
+
+        apps_git.checkout(pr_branch)
+        logging.info("App repo PR branch %s checkout successful", pr_branch)
+        git_hash = apps_git.get_last_commit_hash()
```
Comment (christiansiegel) [inline gitopscli/commands/create_pr_preview.py:42]: We should get the hash from the API as well. It's a bit much overhead to do a full app repo checkout just for this info.

**Pull request merged**